src/public/modules/forms/admin/controllers/view-responses.client.controller.js

```diff
@@ -244,11 +246,102 @@ function ViewResponsesController(
+      params.startDate = moment(new Date(vm.datePicker.date.startDate)).format(
+        'YYYY-MM-DD',
+      )
+      params.endDate = moment(new Date(vm.datePicker.date.endDate)).format(
+        'YYYY-MM-DD',
+      )
+    }
+    Submissions.count(params).then((responsesCount) => {
+      vm.downloadAttachmentResponsesCount = responsesCount
+      vm.confirmDownloadModal = $uibModal.open({
+        backdrop: 'static',
+        templateUrl:
+          'modules/forms/admin/views/download-all-attachments.client.modal.html',
+        scope: $scope,
+      })
+    })
+  }
+
+  vm.downloadMultipleSubmissionAttachments = function () {
+    vm.confirmDownloadModal.close()
+    vm.csvDownloading = true
+    let params = {
+      formId: vm.myform._id,
+      page: -1,
+    }
+    if (vm.datePicker.date.startDate && vm.datePicker.date.endDate) {
+      params.startDate = moment(new Date(vm.datePicker.date.startDate)).format(
+        'YYYY-MM-DD',
+      )
+      params.endDate = moment(new Date(vm.datePicker.date.endDate)).format(
+        'YYYY-MM-DD',
+      )
+    }
+    Submissions.getMetadata(params)
+      .then((data) => {
+        let refIds = data.metadata.map((d) => d.refNo)
+        let downloadPromises = refIds.map((submissionId) => {
+          Submissions.getEncryptedResponse({
+            formId: vm.myform._id,
+            submissionId,
+          }).then((response) => {
+            let displayedContent
+            const { content, verified, attachmentMetadata } = response
+            const decrypted = FormSgSdk.crypto.decrypt(
+              vm.encryptionKey.secretKey,
+              {
+                encryptedContent: content,
+                verifiedContent: verified,
+              },
+            )
+            displayedContent = processDecryptedContent(decrypted)
+
+            let questionCount = 0
+            let attachmentDownloadUrls = new Map()
+            displayedContent.forEach((field) => {
+              // Populate question number
+              if (field.fieldType !== 'section') {
+                field.questionNumber = ++questionCount
+              }
+              // Populate S3 presigned URL for attachments
+              if (attachmentMetadata[field._id]) {
+                attachmentDownloadUrls.set(questionCount, {
+                  url: attachmentMetadata[field._id],
+                  filename: field.answer,
+                })
+                field.downloadUrl = attachmentMetadata[field._id]
+              }
+            })
+
+            return vm.downloadAllAttachments(
+              attachmentDownloadUrls,
+              submissionId,
+            )
+          })
+        })
+        return Promise.all(downloadPromises)
+      })
+      .then(() => {
+        vm.csvDownloading = false
+      })
+  }
+
+  vm.downloadAllAttachments = function (
+    attachmentDownloadUrls = vm.attachmentDownloadUrls,
+    refNo = vm.tableParams.data[vm.currentResponse.index].refNo,
+  ) {
```

Comment:
can we refactor to avoid the circular reference from `vm` to itself?

Reply:
Done.